akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -380,3 +359,142 @@ private[akka] final class SinkQueueAdapter[T](delegate: SinkQueue[T]) extends ak
+  private case class Cancel[T](subscriber: SubscriptionWithCursor[T]) extends PublisherSinkMessages[T]
+  private case class Subscribe[T](subscriber: Subscriber[_ >: T]) extends PublisherSinkMessages[T]
+  private case object CancelPublisher extends PublisherSinkMessages[Nothing]
+
+}
+
+/**
+ * INTERNAL API
+ */
+final private[stream] class CancellablePublisherSink[T](drainWhenNoSubscribers: Boolean) extends GraphStageWithMaterializedValue[SinkShape[T], CancellablePublisher[T]] {
+
+  val in = Inlet[T]("CancellablePublisherSink.in")
+  override def initialAttributes = DefaultAttributes.cancellablePublisherSink
+  override val shape: SinkShape[T] = SinkShape.of(in)
+
+  override def toString: String = "CancellablePublisherSink"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val InputBuffer(initBuffer, maxBuffer) = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+    require(maxBuffer > 0, "Max buffer size must be greater than 0")
+    require(initBuffer > 0, "Init buffer size must be greater than 0")
+
+    val stageLogic = new GraphStageLogic(shape) with CallbackWrapper[PublisherSinkMessages[T]] with SubscriberManagement[T] with InHandler {
+      private var downstreamDemand: Long = 0L
+      private var upstreamCompleted = false
+      private var failedWithError: Throwable = null
+
+      override type S = SubscriptionWithCursor[T]
+
+      override def maxBufferSize: Int = maxBuffer
+      override def initialBufferSize: Int = initBuffer
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        setKeepGoing(true)
+        initCallback(callback.invoke)
+      }
+
+      override def postStop(): Unit = {
+        stopCallback {
+          case CancellablePublisherSink.Subscribe(subscriber) ⇒
+            subscriber.onSubscribe(CancelledSubscription)
+            subscriber.onError(if (failedWithError == null) ActorPublisher.NormalShutdownReason else failedWithError)
+          case _ ⇒ //keep silence
+        }
+      }
+
+      private val callback: AsyncCallback[PublisherSinkMessages[T]] =
+        getAsyncCallback {
+          case CancellablePublisherSink.Subscribe(subscriber)             ⇒ registerSubscriber(subscriber.asInstanceOf[Subscriber[_ >: T]])
+          case CancellablePublisherSink.RequestMore(subscriber, elements) ⇒ moreRequested(subscriber, elements)
+          case CancellablePublisherSink.Cancel(subscription)              ⇒ unregisterSubscription(subscription)
+          case CancellablePublisherSink.CancelPublisher ⇒
+            completeDownstream()
+            completeStage()
+            if (!cursors.isEmpty) {
+              upstreamCompleted = true
+              setKeepGoing(true)
+            }
+        }
+
+      override def onPush(): Unit = {
+        if (cursors.isEmpty) {
+          //drain if no subscriptions
```

Comment:
Sink stops stream at once when last subscriptions been cancelled (synchronously) and strategy is just Fanout (with no Drain). So this code unreachable in case strategy is fanout and all subscriptions are cancelled.
Still let me know if after calling completeStage we still can have Push.